versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/ExportDatabaseAdapter.java

```diff
@@ -97,20 +99,25 @@ private HeadsAndForkPoints scanAllReferences(Consumer<CommitLogEntry> commitHand
           .map(ReferenceInfo::getHash)
           .forEach(
               head -> {
-                try (Stream<CommitLogEntry> commits = databaseAdapter.commitLog(head)) {
-                  for (Iterator<CommitLogEntry> commitIter = commits.iterator();
-                      commitIter.hasNext(); ) {
-                    CommitLogEntry commit = commitIter.next();
-                    if (!identify.handleCommit(commit)) {
-                      break;
-                    }
-                    commitHandler.accept(commit);
+                Deque<Hash> commitsToProcess = new ArrayDeque<>();
+                commitsToProcess.push(head);
+                while (!commitsToProcess.isEmpty()) {
+                  Hash hash = commitsToProcess.pop();
```

Comment:
I guess we may want to check `hash` against the flags in `identify` at this point to avoid accessing the commit log if case the hash has already been processed.

I guess this can be a common situation when two branches have multiple merge points.